test/integration/scheduler/queue_test.go

```diff
@@ -370,6 +370,46 @@ func TestCoreResourceEnqueue(t *testing.T) {
 			},
 			wantRequeuedPods: sets.New("pod1"),
 		},
+		{
+			name: "Pod rejected by the NodeResourcesFit plugin isn't requeued when a Node is updated without increase in the requested resources",
+			initialNodes: []*v1.Node{
+				st.MakeNode().Name("fake-node1").Capacity(map[v1.ResourceName]string{v1.ResourceCPU: "4"}).Obj(),
+				st.MakeNode().Name("fake-node2").Capacity(map[v1.ResourceName]string{v1.ResourceCPU: "2"}).Label("group", "b").Obj(),
+			},
+			pods: []*v1.Pod{
+				// - Pod1 requests available amount of CPU (in fake-node1), but will be rejected by NodeAffinity plugin. Note that the NodeResourceFit plugin will register for QHints because it rejected fake-node2.
+				st.MakePod().Name("pod1").Req(map[v1.ResourceName]string{v1.ResourceCPU: "4"}).NodeAffinityIn("group", []string{"b"}, st.NodeSelectorTypeMatchExpressions).Container("image").Obj(),
+			},
+			triggerFn: func(testCtx *testutils.TestContext) error {
+				// Trigger a NodeUpdate event that increases unrealted (not requested) memory capacity of fake-node1, which should not requeue Pod1.
+				if _, err := testCtx.ClientSet.CoreV1().Nodes().UpdateStatus(testCtx.Ctx, st.MakeNode().Name("fake-node1").Capacity(map[v1.ResourceName]string{v1.ResourceCPU: "4", v1.ResourceMemory: "4000"}).Obj(), metav1.UpdateOptions{}); err != nil {
```

Comment:
We have the same problem here: https://github.com/kubernetes/kubernetes/pull/127444#discussion_r1773398189

Reply:
Wouldn't it eventually succeed?
Because there is a call to `PollUntilContextTimeout` after calling `triggerFn`.

Reply:
I'm trying to understand the problem. Is it a problem with any Update call in this test case, or in this specific case?

Reply:
It's a good catch.



So, the problem is that we cannot know whether the Node/Update hasn't been handled yet or it's been handled and not requeued the Pod as a result of QHints.

Reply:
Actually this test is growing, and I think we should prioritize the solution here. Calling `SchedulingQueue.Update` is every test isn't ideal.

Can we add a metric `queueing_hint_evaluation_total` (like [plugin_evaluation_total](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/metrics/metrics.go#L312-L318)) and ensure the QHint is called (= the event is arrived and handled) expectedly after `triggerFn`?
Or, if it's not worthy enough to be a metric, we can add a counter in [the struct](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/backend/queue/scheduling_queue.go#L152), representing the number of events the scheduling queue handles.

Reply:
> I'm trying to understand the problem. Is it a problem with any Update call in this test case, or in this specific case?

It's a problem with only the cases for "not requeue", i.e. `wantRequeuedPods: sets.Set[string]{}`

> Wouldn't it eventually succeed?
> Because there is a call to PollUntilContextTimeout after calling triggerFn.

Yes, It will succeed eventually, but it could also succeed before the scheduler has handled the event:

1. Pods are stayed in the unscheduled queue initially.
2. Node was updated.
3. Scheduler began to handle the node/update event.
4. Meanwhile, `PollUntilContextTimeout` is called, we still got an empty activeQ here, so the test case returns success immediately.
5. QHint actually starts now, and put the pods into the activeQ, but we missed this.

Reply:
Ah, wait. Rather than adding a new metric, maybe we can take a number of samples from `event_handling_duration_seconds` somehow.
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/metrics/metrics.go#L143-L151

Reply:
https://github.com/kubernetes/kubernetes/pull/127607

Reply:
> It's a problem with only the cases for "not requeue", i.e. `wantRequeuedPods: sets.Set[string]{}`

I see. In that case, this suggestion is about improving the test, but if there was a bug in the re-queuing logic, the test would start flaking, giving us enough signal.
In that case, I'm happy with the current implementation and to leave the rest for a follow up such as #127607

Reply:
Great, it should resolve the problem. I will comment in https://github.com/kubernetes/kubernetes/pull/127607

Reply:
I'm not sure whether it can be really a problem, but I found checks of pendingPods before and after ScheduleOne calls at least confusing. pendingPods number is a sum of active, backoff and unschedulable queues, where IIUC backoff queue is in fact always empty (uses zero backoff seconds).

In https://github.com/kubernetes/kubernetes/pull/127641 I added a check that the active queue is really empty after scheduling all Pods, which should imply the Pods are indeed in unschedulable queue. It should at least improve readability of the test by asserting the assumptions.